plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,457 @@
+
+// ContainsService indicates whether a particular service is enabled
+func (s serviceMap) ContainsService(t serviceType) bool {
+	for _, service := range s {
+		if service.Type == string(t) {
+			return true
+		}
+	}
+	return false
+}
+
+// projectMap maps a project id to a Project struct.
+type projectMap map[string]projects.Project
+
+// hypervisorMap maps a hypervisor id a Hypervisor struct.
+type hypervisorMap map[int]hypervisors.Hypervisor
+
+// serverMap maps a server id to a Server struct.
+type serverMap map[string]servers.Server
+
+// flavorMap maps a flavor id to a Flavor struct.
+type flavorMap map[string]flavors.Flavor
+
+// volume is a structure used to unmarshal raw JSON from the API into.
+type volume struct {
+	volumes.Volume
+	volumetenants.VolumeTenantExt
+}
+
+// volumeMap maps a volume id to a volume struct.
+type volumeMap map[string]volume
+
+// storagePoolMap maps a storage pool name to a StoragePool struct.
+type storagePoolMap map[string]schedulerstats.StoragePool
+
+// OpenStack is the main structure associated with a collection instance.
+type OpenStack struct {
+	// Configuration variables
+	IdentityEndpoint string
+	Domain           string
+	Project          string
+	Username         string
+	Password         string
+
+	// Locally cached clients
+	identity *gophercloud.ServiceClient
+	compute  *gophercloud.ServiceClient
+	volume   *gophercloud.ServiceClient
+
+	// Locally cached resources
+	services     serviceMap
+	projects     projectMap
+	hypervisors  hypervisorMap
+	flavors      flavorMap
+	servers      serverMap
+	volumes      volumeMap
+	storagePools storagePoolMap
+}
+
+// Description returns a description string of the input plugin and implements
+// the Input interface.
+func (o *OpenStack) Description() string {
+	return "Collects performance metrics from OpenStack services"
+}
+
+// sampleConfig is a sample configuration file entry.
+var sampleConfig = `
+  ## This is the recommended interval to poll.
+  interval = '60m'
+
+  ## The identity endpoint to authenticate against and get the
+  ## service catalog from
+  identity_endpoint = "https://my.openstack.cloud:5000"
+
+  ## The domain to authenticate against when using a V3
+  ## identity endpoint.  Defaults to 'default'
+  domain = "default"
+
+  ## The project to authenticate as
+  project = "admin"
+
+  ## The user to authenticate as, must have admin rights
+  username = "admin"
+
+  ## The user's password to authenticate with
+  password = "Passw0rd"
+`
+
+// SampleConfig return a sample configuration file for auto-generation and
+// implements the Input interface.
+func (o *OpenStack) SampleConfig() string {
+	return sampleConfig
+}
+
+// gather is a wrapper around library calls out to gophercloud that catches
+// and recovers from panics.  Evidently if things like volumes don't exist
+// then it will go down in flames.
+func gather(f func() error) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
```

Comment:
Don't handle this, actually Telegraf does this in the agent for you, though in the future I may remove this since it gives very little recoverability what with user started goroutines.  Best IMO is just not to panic and only recover if you know a panic is expected.

Reply:
I've removed this wrapper function.